### DISCUSSION TITLE
FontAwesome CDN

### DISCUSSION THREAD

**Discussion by anytizer** [opencart/opencart#10947] [General]
https://github.com/opencart/opencart/tree/master/upload/install/view/stylesheet/fontawesome

May be it is a time to upgrade FontAwesome into CDN.
Looks like only install/ has it.

**Comment by felipoantonoff** [comment]
It's an interesting idea, I've been using a direct CDN like https://cdnjs.com for a few years, there's only one detail that has a CDN that sometimes gets unstable, since these larger ones usually have more than one CDN to serve.

Perhaps an idea in the future would be to analyze which public CDN is better, exchange the codes directly for them or in the Panel even the customer may be able to choose between which public CDN he wants.

I've been using it for years on the Panel and on the company's main website, I prefer as soon as I compile JS or have to take care of the files, but I also understand the use of CSS and JS in place, as it creates 0 external dependency and many Stores use their own CDN .

Another way is to give the client the option to use it by CDN or local, maybe even at the time of installation or after the configuration it can change.

**Comment by stalker780** [comment]
Does not matter where they are loaded from.

I am more interested, how to remove unneeded default external resources. For example I am building a theme where they are not used, fonts, javascripts and styles.

OC4 have no function to check and remove links, styles, javascripts using events.

OC4 has only Add and Get functions. But not Delete.
https://github.com/opencart/opencart/blob/b9e0499caadcd2f2b8c0ddbdc12b551c58e91e16/upload/system/library/document.php#L84

And there is no way to remove unused resources with events. Only parse twig file, or it's data.

**Reply by felipoantonoff** [reply]
The way to remove it for certain cases is interesting, but maybe it's only possible easily for JS and CSS defined in the Controller, since the direct ones in Twig I imagine only by OCMOD (XML) or with Events doing all that work of getting the original content of the view , str_replace and the like and return the modified view, not to mention that it can conflict, so I still don't see any sense in Events for this type of use, but for now without OCMOD it's the only way I can imagine.

**Comment by anytizer** [comment]
It might be used **temporarily only**, because we can see entire FA inside install/ folder. OC itself suggests to remove the Install folder once the installation is over.

**Comment by osworx** [comment]
Voted "down".

Why?
Loading external resources is not the best idea.

2 reasons:
1. avoid to leave leads somewhere (in the sence of Privacy and other)
2. when working in a "locale" environement (e.g. during development) some may have no Internet (or a very slow one - yes this happens daily and anywhere!)

Better approach would be, to load every library/font/etc. from locale instead fetching external sources.
Beside this, in many cases this is faster (testes that in several installations and have a 30-60% better loading performance).

**Reply by felipoantonoff** [reply]
Makes sense, but maybe it could be an option primarily for production.

CDN in theory has better performance, but it can actually be below Hosting in certain cases, especially if it is already in the customer's region.

Now this change is really not a priority, because the CDN can be put in the Store in general later if necessary.

**Comment by anytizer** [comment]
Agreeable. But at least, there could have been single instance of such content.

For example:
- https://github.com/opencart/opencart/tree/master/upload/install/view/stylesheet/fontawesome
- https://github.com/opencart/opencart/tree/master/upload/admin/view/stylesheet/fonts/fontawesome
- https://github.com/opencart/opencart/tree/master/upload/catalog/view/stylesheet/fonts/fontawesome

**Reply by felipoantonoff** [reply]
Really good to keep only one, I've even seen JS give conflict, because modules climbed the same JS.

The CSS and JS could be in a more general assets folder, at least those used by the platform's core, it could even be inside the storage / I imagine.

If it's CSS or JS just for installation, it would be interesting to CDN to reduce the amount of CSS and JS.
